### PR TITLE
feat(CDCL): Add ability to decide on semantic literals

### PR DESCRIPTION
**Pull request by bclement-ocp**
This patch introduces the ability for the CDCL solver to deal with semantic literals. This is part of work towards #901 although it does not (yet) achieves the goals of #901.

More precisely:

 - We add a new `Literal` module in `structures` to represent literals that are either syntaxic (represented by expressions) or semantic (represented using the `Xliteral` module).

 - The new `Literal` module is used to represent the literals associated with the `Atom`s in the CDCL solver [1]

 - The CDCL solver has a special handling of these semantic literals, which are treated as "local" literals: they are not added in the variable ordering, and will not be decided upon by the solver. Instead, a new mechanism inspired by mSAT (and, transitively, mcSAT) is provided, where the theory is able to mark certain literals as "immediate decisions". When literals are marked this way, the solver will decide on one of the marked literals the next time a decision must be made. Literals marked as immediate decisions are part of a transient solver state, and are forgotten upon backtracking.

 - The theory uses this marking mechanism to communicate case splits (and a similar mechanism for optimization objectives, which is separate because the value of the objective is sent back to the theory once the decision is made) to the solver.

This mechanism allows to introduce splits that are, in a way, "local", and do not  pollute other branches of the solver. It can be seen as an intermediate step towards a mcSAT approach, but where we are still deciding on boolean literals and hence try to avoid adding a large number of splits globally into the solver -- although it does mean that the splits are not incorporated into the activity heuristic.

The functionality is currently gated behind a new flag `--enable-sat-cs`.

Note that with `--enable-sat-cs` this change is +46-13 on ae-format (and +8-7 on benchtop-tests) or a net positive of +33 on ae-format, so we might want to enable it by default.

[1] : It could also be used for the `Bj`s in the Tableaux solver, if we
      wanted to suport similar features in the Tableaux solver.

### CODE REVIEW

**Comment by bclement-ocp** [comment]
I forgot to mention but: the `--enable-sat-cs` flag currently only applies to the Arrays theory, because for some theories (e.g. integers) it leads to regressions, probably due to polluting the main environment too much with irrelevant decisions. Further work towards #901 should investigate strategies to gradually move the other theories to this mechanism without introducing regressions (the enum and bit-vector theories should be prime candidates).

File: src/lib/structures/literal.ml
Comment (Halbaroth) [inline src/lib/structures/literal.ml:85]: Ah ah :D I know exactly where you got this comment and I don't know what it means too.

File: src/lib/structures/literal.mli
Comment (Halbaroth) [inline src/lib/structures/literal.mli]: As `l'` is the returned value of `normal_form l`, it's easier to understand these conditions in terms of `l`:
```suggestion
      - [l'] is [neg l] if [is_neg] is [true]
      - [l'] is [l] if [is_neg] is [false]
      - [normal_form l'] is [l', false] *)
```
and `is_neg = is_neg l`.

File: src/lib/structures/objective.mli
Comment (Halbaroth) [inline src/lib/structures/objective.mli:94]: We have to update the documentation too ;)

File: src/lib/structures/satml_types.mli
Comment (Halbaroth) [inline src/lib/structures/satml_types.mli]: I'm not very fond of this name. We may call it `add_sem_atom`?

File: src/lib/reasoners/satml.ml
Comment (Halbaroth) [inline src/lib/reasoners/satml.ml:395]: We may say that these decisions are always semantic.

File: src/lib/reasoners/satml_frontend.ml
Comment (Halbaroth) [inline src/lib/reasoners/satml_frontend.ml]: Is it an oversight or this case can occur?

File: src/lib/reasoners/theory.ml
Comment (Halbaroth) [inline src/lib/reasoners/theory.ml]: :D I didn't open an PR for such a small mistake 

```suggestion
           expression is seen as uninterpreted term of the arithmetic theory.
```

File: src/lib/reasoners/theory.ml
Comment (Halbaroth) [inline src/lib/reasoners/theory.ml]: We may mention here the cause of this issue, that is some semantic literals can be repushed on the trail of SatML after backtracking the environment of the union-find ;)

File: src/lib/reasoners/theory.ml
Comment (Halbaroth) [inline src/lib/reasoners/theory.ml:757]: It seems that `assumed_set` is used for two things:
- it's a field of the record `Models.t` which is used in a deprecated feature.
- If we use a tableaux-like method for theories (which is the default), we try to instantiate free variables of the atoms of `assumed_set` in order to produce new facts. I guess the idea is to reach an eventual contradiction by using the instantiation engine here. With the current implementation, we cannot match semantic values, so it makes sense you ignore semantic literals here, but I wonder if it would be relevant if we could do it? I believe it's not relevant because such contradictions should be found in the relation theories before emitting the case-split.

File: src/lib/reasoners/intervalCalculus.ml
Comment (Halbaroth) [inline src/lib/reasoners/intervalCalculus.ml:2078]: We should update the documentation of `middle_value` as we don't use this function anymore to force a specific value.

File: src/lib/reasoners/intervalCalculus.ml
Comment (Halbaroth) [inline src/lib/reasoners/intervalCalculus.ml]:
```suggestion
            [r1; middle_value env ~is_max ty p None],
```

**Comment by bclement-ocp** [comment]
So it turns out that I forgot to reset the number of choices when backtracking, which means that we reply "sat" from satML too early, which in turns means we are more aggressive with instantiation, and that seems to be beneficial somehow, since we get an improvement… but doing what should be the right thing (i.e. actually counting semantic decisions properly) causes large regressions (`+69-137`). I will investigate next week.

**Comment by bclement-ocp** [comment]
> So it turns out that I forgot to reset the number of choices when backtracking, which means that we reply "sat" from satML too early, which in turns means we are more aggressive with instantiation, and that seems to be beneficial somehow, since we get an improvement… but doing what should be the right thing (i.e. actually counting semantic decisions properly) causes large regressions (`+69-137`). I will investigate next week.

Turns out I made a typo while rebasing: https://github.com/OCamlPro/alt-ergo/pull/1027/commits/d708f3315ca9b0bc250c90d5f81160cf85724905

Rerunning the benches.

**Comment by bclement-ocp** [comment]
The benches now say `+49-14` with the correct code rebased on e18c2d17

File: src/lib/reasoners/th_util.mli
Comment (Halbaroth) [inline src/lib/reasoners/th_util.mli]:
```suggestion
      [SAT]. The order of decisions among multiple calls of
      [acts_add_decision_lit] is unspecified.
```

File: src/lib/reasoners/theory.ml
Comment (Halbaroth) [inline src/lib/reasoners/theory.ml]:
```suggestion
    (* We have to add the term [fn] and its subterms as the MaxSMT
       syntax allows to optimize expressions that aren't part of
       the current context. *)
```

File: src/lib/reasoners/theory.ml
Comment (Halbaroth) [inline src/lib/reasoners/theory.ml:518]: After this modification, I believe the `aux` dispatcher isn't useful anymore. If `generate_choices` found new choices, we can call `propagate_choices` directly in this function. In `propagate_choices`, we call `generate_choices` if there is no more choices to propagate. Finally we replace the top-level call of `aux` in `look_for_sat` by:
```ocaml
propagate_choices env sem_facts (List.rev env.choices) new_choices
```

Notice we mention `aux` in the documentation of this function too ;)

File: src/lib/reasoners/theory.ml
Comment (Halbaroth) [inline src/lib/reasoners/theory.ml:623]: From my understanding, this piece of code ensures we don't reuse previous choices in presence of optimization after backtracking in the SAT solver. Is it correct to remove it because we entail optimization splits in the SAT environment in this PR?

File: src/lib/reasoners/satml.ml
Comment (Halbaroth) [inline src/lib/reasoners/satml.ml]: I'm not sure to understand this comment. Do you avoid semantic literals here for sake of soundness or efficiency? From my understanding, if we learn something in the CDCL framework, this fact have to be entailed by the problem. It may be irrelevant after backtracking but still correct. So I guess you meant we should avoid learning semantic facts because the case-split mechanism could produce a lot of them and we don't want to increase too much the size of the clause database?

**Review by Halbaroth — APPROVED** [review]

File: src/lib/structures/objective.mli
Comment (bclement-ocp) [reply src/lib/structures/objective.mli:94]: Done.

File: src/lib/structures/satml_types.mli
Comment (bclement-ocp) [reply src/lib/structures/satml_types.mli]: I don't like it either, but I do not have a better solution. `add_sem_atom` would be confusing IMO because `Shostak.Literal.t` can contain either "syntaxic" literals (`LTerm`) or "semantic" literals (`LSem`).

Maybe renaming this one to `add_atom` and the current `add_atom` to `add_expr_atom`?

File: src/lib/reasoners/satml.ml
Comment (bclement-ocp) [reply src/lib/reasoners/satml.ml:395]: I'd rather not; while it is true that currently only semantic decisions are added here, the implementation doesn't actually care.

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: Oversight; fixed.

File: src/lib/reasoners/satml.ml
Comment (bclement-ocp) [reply src/lib/reasoners/satml.ml]: Clauses are learned at level 0 but semantic literals can includes terms created by the Shostak module that depend on assertions at greater levels, and hence do not exist (in the union-find etc.) at level 0 (what I mean by "no longer valid" here). I have updated and clarified the comment.

File: src/lib/reasoners/theory.ml
Comment (bclement-ocp) [reply src/lib/reasoners/theory.ml:518]: Done.

File: src/lib/reasoners/theory.ml
Comment (bclement-ocp) [reply src/lib/reasoners/theory.ml:623]: Yes, now that optimizations are applied in the main environment, nothing special needs to be done in the case-split environment, it just fits the existing framework.

File: src/lib/reasoners/theory.ml
Comment (bclement-ocp) [reply src/lib/reasoners/theory.ml]: Clarified the comment.

File: src/lib/reasoners/theory.ml
Comment (bclement-ocp) [reply src/lib/reasoners/theory.ml:757]: For the first point: the `prop_model` in model generation needs the value of the boolean variables of the original problem; semantic literals are not in the original problem and are not needed there.

For the second point: theoretically I guess we could, however I believe that the `assumed_set` is only used to expand user-defined predicates, so it would not really be useful here.

File: src/lib/reasoners/intervalCalculus.ml
Comment (bclement-ocp) [reply src/lib/reasoners/intervalCalculus.ml:2078]: Done.

**Review by bclement-ocp — COMMENTED** [review]
Should be ready for round 2.

(Note: the changes are fairly involved and so I want to re-run the benches with the exact code that will get merged prior to merging; please don't merge when approving)

**Comment by Halbaroth** [comment]
Looks good to me, you can merge the PR.

**Comment by bclement-ocp** [comment]
There is an additional bit of code that I believe is needed in `try_to_backjump_further` (we could end up taking a semantic decision from the "fake" `lazy_cnf` that we shouldn't have taken); I will fix this and rerun benchmarks before merging.

**Comment by bclement-ocp** [comment]
> There is an additional bit of code that I believe is needed in `try_to_backjump_further` (we could end up taking a semantic decision from the "fake" `lazy_cnf` that we shouldn't have taken); I will fix this and rerun benchmarks before merging.

Pushed the aforementioned change as well as separation of `next_decisions` into `next_decisions` (mandatory, keep all) and `next_split` (optional, keep only one). This is because otherwise in the initial processing phase (before actually solving) we will add the same split over and over again to `next_decisions`. I have kept `next_decisions` even though it is no longer used; I'm not too sure about this — it can be useful in the near future but also could be re-introduced at that point.

Re-running benchmarks before merging; although we still need to decide if `--enable-sat-cs` should be on by default (I'd say yes, but I'm also OK waiting until we can enable it for all theories w/o regressions).

**Comment by bclement-ocp** [comment]
The current version is +50-8 on AE-format.

**Comment by bclement-ocp** [comment]
I will go ahead and merge this now with the flag disabled — I think it is acceptable to have it as an experimental flag (and this simplifies the optimization code somewhat) but we should have a comprehensive story for all theories before enabling the flag by default.

**Pull request merged**
